### PR TITLE
VideoCheckers Praktikum KI 2025

### PR DESCRIPTION
**Pull request by Baspla**

### CODE REVIEW

**Comment by Baspla** [comment]
I got a notification that the test run was "Cancelled after 29s".
This sounds like a manual step by someone?
Is there anything actionable to do on out side?

**Comment by Baspla** [comment]
From the logs i assume there is something wrong with our observations?

**Comment by Baspla** [comment]
I don't think ill get anywhere here today.
Is there any documentation on how to do the observation_space method?

**Comment by PaulSeitz** [comment]
> I don't think ill get anywhere here today. Is there any documentation on how to do the observation_space method?

I think your problem is not the observation_space method but rather that your reset and step method return different formats? 
Quickly going over your code I think the culprit might be your manual frame stacking in the reset function?
`        
def expand_and_copy(x):
            x_expanded = jnp.expand_dims(x, axis=0)
            return jnp.concatenate([x_expanded] * self.frame_stack_size, axis=0)

        # Apply transformation to each leaf in the pytree
        initial_obs = jax.tree.map(expand_and_copy, initial_obs)
`

**Comment by Baspla** [comment]
I mean we have been working on this since the beginning of the semester and the process was "look at the examples" so we copied the base structure from game to game.

**Comment by PaulSeitz** [comment]
> I mean we have been working on this since the beginning of the semester and the process was "look at the examples" so we copied the base structure from game to game.

Yes that would be legacy code that was overlooked from our side
